hnn_core/gui/gui.py

```diff
@@ -504,129 +921,162 @@ def load_drives(variables, params, log_out, drives_out, drive_widgets,
             )
 
 
-def on_upload_change(change, sliders, params, tstop, tstep, log_out, variables,
-                     drive_boxes, drive_widgets, drives_out):
+def on_upload_change(change, params, tstop, tstep, log_out, variables,
+                     drive_boxes, drive_widgets, drives_out, connectivity_out,
+                     connectivity_sliders, load_info):
     if len(change['owner'].value) == 0:
         return
 
-    params_fname = change['owner'].metadata[0]['name']
+    # params_fname = change['owner'].metadata[0]['name']
+    params_fname = list(change['owner'].value.keys())[0]
     file_uploaded = change['owner'].value
     param_data = list(file_uploaded.values())[0]['content']
     param_data = codecs.decode(param_data, encoding="utf-8")
 
+    if load_info['prev_param_data'] == param_data:
+        with log_out:
+            print("Same param. No reloading")
+        return
+    else:
+        load_info['prev_param_data'] = param_data
+
     ext = op.splitext(params_fname)[1]
     read_func = {'.json': _read_json, '.param': _read_legacy_params}
     params_network = read_func[ext](param_data)
 
+    # update simulation settings and params
     log_out.clear_output()
     with log_out:
-        print(f"parameter key: {params_network.keys()}")
-        for slider in sliders:
-            for sl in slider:
-                key = 'gbar_' + sl.description
-                sl.value = params_network[key]
-
         if 'tstop' in params_network.keys():
             tstop.value = params_network['tstop']
         if 'dt' in params_network.keys():
             tstep.value = params_network['dt']
 
         params.update(params_network)
-    load_drives(variables, params, log_out, drives_out, drive_widgets,
-                drive_boxes, tstop)
+    # init network, add drives & connectivity
+    load_drive_and_connectivity(variables, params, log_out, drives_out,
+                                drive_widgets, drive_boxes, connectivity_out,
+                                connectivity_sliders, tstop, load_info)
+
+
+def _init_network_from_widgets(params, tstep, tstop, variables, drive_widgets,
+                               connectivity_sliders, add_drive=True):
+    """Construct network and add drives."""
+    print("init network")
+    params['dt'] = tstep.value
+    params['tstop'] = tstop.value
+    variables['net'] = jones_2009_model(
+        params,
+        add_drives_from_params=False,
+    )
+    # adjust connectivity according to the connectivity_tab
+    for connectivity_slider in connectivity_sliders:
+        for vbox in connectivity_slider:
+            conn_indices = pick_connection(
+                net=variables['net'],
+                src_gids=vbox._belongsto['src_gids'],
+                target_gids=vbox._belongsto['target_gids'],
+                loc=vbox._belongsto['location'],
+                receptor=vbox._belongsto['receptor'])
+
+            if len(conn_indices) > 0:
```

Comment:
There should be a check to ensure that there is only one element in `conn_indices`